### ISSUE TITLE
Add `text-wrap: pretty;` as a global toggle option

### ISSUE THREAD

**Issue by jasmussen** [WordPress/gutenberg#55190]
The CSS feature `text-wrap: pretty;` enables line-breaks to happen in a way that avoids typographic widows. You can [read a bit more here](https://clagnut.com/blog/2424/), although that refers to `balance` where `pretty` works slightly differently. The example from that post, shown here, features a typographic widow:

```
One year on and what next for remote
working?
```

The CSS property "balanced" would enable that heading to wrap instead like this:

```
One year on and what
next for remote working?
```

The other one, "pretty", would wrap it specifically to avoid the typographic widow:

```
One year on and what next for
remote working?
```

The feature is useful for headings and captions especially, and it would be nice if there was a way to apply this in the block editor. 

## Proposed solution

In global styles, as part of the typography panel, add options to adjust the text-balance:

<img width="1200" alt="Text-Wrap i4" src="https://github.com/user-attachments/assets/a2c2a001-a219-4539-b5be-9f940f45b89b">

These properties should ideally show up _only_ in Global Styles, as they are not useful on a per-block basis.

**Note: ** `text-wrap: pretty;` still isn't supported in [safari & firefox](https://caniuse.com/?search=text-wrap%20pretty). So if that option is chosen, it should should add `text-wrap: balance;` as a fallback.

See https://github.com/WordPress/gutenberg/pull/62233 for an example

Issue updated Aug 13.

**Comment by richtabor** [comment]
Related in part: https://github.com/WordPress/gutenberg/issues/48202

**Comment by simison** [comment]
In the editor you might want to have "stable" value for things that are editable, if I understood correctly:

> [...] [pretty](https://www.w3.org/TR/css-text-4/#valdef-text-wrap-style-pretty) is offered as an opt-in to take more time for better results. The pretty value is intended for body text, where the last line is expected to be a bit shorter than the average line; the [balance](https://www.w3.org/TR/css-text-4/#valdef-text-wrap-style-balance) value is intended for titles and captions, where equal-length lines of text tend to be preferred; **and the [stable](https://www.w3.org/TR/css-text-4/#valdef-text-wrap-style-stable) is intended for sections that are, or are likely become toggled as, editable.**

— [via](https://www.w3.org/TR/css-text-4/#ref-for-valdef-text-wrap-style-auto%E2%91%A4)

**Comment by jasmussen** [comment]
Indeed, balance seems too balanced in cases, and _pretty_ seems even better.

**Comment by hanneslsm** [comment]
Somehow related: While `text-wrap: pretty` is something worth exploring, adding manually line-breaks would be also very beneficial: https://github.com/WordPress/gutenberg/issues/55565 Both features could also work well together.

**Comment by richtabor** [comment]
I've found a combination of both to work nicely:

```
/* Progresive enhancement to reduce widows and orphans. */
h1, h2, h3, h4, h5, h6, blockquote, caption, figcaption {
    text-wrap: balance;
}

p {
    text-wrap: pretty;
}
```

**Comment by jasmussen** [comment]
Might be good to look at this as a group; typography presets, spacing, text indents, text-wrapping, etc. Some of these properties apply differently to headings vs. paragraphs, etc. So could use a broader look.

**Comment by MaggieCabrera** [comment]
Would this be something we ever surface as a control in the UI too or just theme.json?

**Comment by jasmussen** [comment]
> Would this be something we ever surface as a control in the UI too or just theme.json?

I think we should always strive to have UI options for each theme.json property, with only a few exceptions. Though it can certainly start in theme.json only, and get a UI at a later time. There's also a question on whether we want options, i.e. do you want it for headings only or also for paragraphs? Do you want _pretty_ or _balance_? Perhaps _nowrap_ and _wrap_?

It could be something like this:

<img width="640" alt="Text-Wrap" src="https://github.com/WordPress/gutenberg/assets/1204802/17bbb730-8cec-4478-b0b3-99d0ddcba949">


[Figma](https://www.figma.com/design/BAssQhQUVLPE1ld42m2Dn6/Global-Styles-%26-Saving-Style-Variations?node-id=1-2638&t=lpiBdnk4y7AVs7vf-1).

**Comment by jasmussen** [comment]
A note to self: these text wrapping properties are related to other global writing settings like text-justification (justified text etc) that make sense _only when they are global_.

**Comment by benniledl** [comment]
Please also add the ability to add an &shy; inline element so that languages where a word needs to break at a specific spot can do so.
https://github.com/WordPress/gutenberg/issues/55565

**Comment by jasmussen** [comment]
@at-benni can you use shift-enter to force an inline line-break?

**Comment by benniledl** [comment]
It should not break on desktop, it should only break when it does not have enough space. Just like `&shy;` does

**Comment by jasmussen** [comment]
I wasn't familiar with `&shy;`, looks like it's a soft-hyphen. But I'm not sure how you'd use that. Would something like [CSS hyphenation](https://developer.mozilla.org/en-US/docs/Web/CSS/hyphens) options work?

**Comment by benniledl** [comment]
I am not sure, I tested it but I don't think it's suitable.

German has a lot of long words, and provides rules where the word break should be so it should not just be random but it should be decided by the editor.

Arbeiterunfallversicherungsgesetz should break like this: 
Arbeiterunfall-
versicherungsgesetz

**Comment by fabiankaegy** [comment]
Also just noting that `text-wrap: pretty` still isn't supported in [safari & firefox](https://caniuse.com/?search=text-wrap%20pretty). So we should add `text-wrap: balance` as a fallback.

See https://github.com/WordPress/gutenberg/pull/62233 for an example

**Comment by jasmussen** [comment]
Here's another somewhat simpler mockup:

<img width="640" alt="Text-Wrap i2" src="https://github.com/user-attachments/assets/28731fc9-2e78-4bec-8626-b62fbf68e8bb">

In this one, pretty text-wrapping simply becomes a toggle. There's room in this mockup also for future text-indent toggles, as well as orientation. The latter may or may not be a fit for global styles, but for now I've made room for it. @richtabor any thoughts? It seems a fairly simple thing to ask for dev-help on, if we can unstick the design. CC: @jameskoster as well for instincts.

**Comment by jameskoster** [comment]
Personally I prefer the previous mockup. In this latest concept I worry too much information has been extracted, to the point it's not clear what the option does. _Maybe_ it would work with different icons, but to me this feels like a case where some help text is warranted.

**Comment by jasmussen** [comment]
Here's a new mockup that restores the dropdown. It can then contain both "Pretty" and "Balance", where the latter can be good for some heading-based designs.

<img width="1200" alt="Text-Wrap i3" src="https://github.com/user-attachments/assets/03e787a7-d866-4ee2-908c-feadbd8f6263">

Edit: updated image to fix "appearance" typo.

This uses the same dropdown type that "Appearance" does, allowing it to show help-text next to each item.

**Comment by jameskoster** [comment]
Assuming the label should be "Text wrap" rather than "Appearance" that looks good to me, outside of a couple of minor copy considerations:

* Unsure if Default needs help text.
* "No wrapping" → "None"?
* Shorter help text for pretty: "Avoid typographic widows"?
* Shorter help text for balanced: "Evenly distribute text across lines"?
* Shorter no wrapping help text: "Force full text on to a single line"?

**Comment by jasmussen** [comment]
Good catch on the typo, yes it should be text wrap. Trying an i4 based on your feedback:

<img width="1200" alt="Text-Wrap i4" src="https://github.com/user-attachments/assets/a2c2a001-a219-4539-b5be-9f940f45b89b">

Great text-change suggestions. I did keep the help-text for default, though. Both because the 4 options look more balanced and equal with that in place, also because what actually _is_ "default"?

**Comment by jameskoster** [comment]
:shipit:

**Comment by jasmussen** [comment]
Updated the issue and shuffled the labels to indicate, this is ready!

**Comment by Drivingralle** [comment]
I like the idea.
My suggestion to make the effort of adding another control to the interface even more valuable, to also add the setting of "nowrap".

That can be very useful while creating websites with lists or tables full of details that should not be wrapped. like pricing tables.

**Comment by hanneslsm** [comment]
> even more valuable, to also add the setting of "nowrap".

nowrap is for the average user probably only useful when there are also overflow settings. Could getting very complicated though.

**Comment by jasmussen** [comment]
> My suggestion to make the effort of adding another control to the interface even more valuable, to also add the setting of "nowrap".

Nowrap is included as the "None" option, with additional help-text below.

**Comment by davewhitley** [comment]
In my experience, I wouldn't want text wrap balance as a fallback because it has major effects on the wrapping and the design of the page. I'd rather have text wrap pretty with no fallback.

**Comment by mcatanzaro** [comment]
Relevant blog post: [Better typography with text-wrap pretty](https://webkit.org/blog/16547/better-typography-with-text-wrap-pretty/)

> In my experience, I wouldn't want text wrap balance as a fallback because it has major effects on the wrapping and the design of the page. I'd rather have text wrap pretty with no fallback.

There's an example in that blog demonstrating why `text-wrap: balance` is probably not a good idea:

![Image](https://github.com/user-attachments/assets/6619b69c-0d31-471f-889b-e167d9e7e2b8)

**Comment by benoitchantre** [comment]
[Better typography with text-wrap pretty](https://webkit.org/blog/16547/better-typography-with-text-wrap-pretty/), a great post written by Jen Simmons on webkit.org

**Comment by mcatanzaro** [comment]
Anyway, I'm sort of in the wrong issue tracker here, but https://github.com/WordPress/twentytwentyfive has been archived and replaced with a link to https://core.trac.wordpress.org/search?q=twenty+twenty-five which is broken (it's a 403 Forbidden page), so... I guess I'll just say beware, as there can be unexpected consequences to globally enabling `text-wrap: pretty`. I found my way here from [this stylesheet comment](https://github.com/WordPress/twentytwentyfive/blob/62c10821880ac2242460e7ecf42727ce1a026047/style.css#L49):

```
/*
 * Progressive enhancement to reduce widows and orphans
 * https://github.com/WordPress/gutenberg/issues/55190
 */
h1, h2, h3, h4, h5, h6, blockquote, caption, figcaption, p {
	text-wrap: pretty;
}
```

Here's a screenshot of my personal blog using stable WebKitGTK 2.48.0, which does not support `text-wrap: pretty`. This is using the TwentyTwentyFive theme with Blog Home template set to "Text-only blog, home":

![Image](https://github.com/user-attachments/assets/14153fb6-df04-4536-93df-9137938e86cd)

And here is same page using newer WebKitGTK 2.49.1, which supports `text-wrap: pretty`:

![Image](https://github.com/user-attachments/assets/f9e48019-832a-4bdc-9165-bcf947dc6f5a)

Notably, the Gutenberg editor always displays the former (good) version even with newer WebKit. I think the theme's global stylesheet is only used on the live website and not by the Gutenberg editor itself. Evidently the theme's global CSS is a risky place to add styles?

I'm not expecting tech support here in the Gutenberg issue tracker, and I don't think I can fix it anyway, since my Wordpress instance doesn't allow custom CSS, and the various layout options are not sufficient. Just posting this as a warning, something to consider when adding this toggle option....

**Comment by andreawetzel** [comment]
I really like the proposed solution to add `text-wrap: pretty` or `text-wrap: balance` at the block level. However, it would be better for our sites if `text-wrap:nowrap` could be added inline from the Rich Text editor. 

For example to prevent a hyphenated name from breaking where we don't wish: `<span style="text-wrap:nowrap">UW–Madison</span>` but this will still allow the rest of the text to wrap as expected. [Here is a Codepen to help illustrate this as an inline style.](https://codepen.io/andreawetzel/pen/wBKoXRN/a427b3687dbec5a6626ce47a7da60b5e?editors=1100)

**Comment by t-hamano** [comment]
Note that while `text-wrap: balance` may work well in Western languages, it may cause unintended wrapping in other languages, for example in Japanese, `text-wrap: balance` produces unnatural wrapping. Please refer to the discussion on https://github.com/WordPress/wporg-parent-2021/pull/171. Furthermore, `text-wrap:pretty` has no effect on Japanese.

Personally, I think this feature should be disabled by default, even if a UI is added.

**Comment by davewhitley** [comment]
related: #64000

**Comment by mtias** [comment]
Off by default seems fine.
